src/content/en/updates/2019/02/hotpath-with-wasm.md

```diff
@@ -0,0 +1,492 @@
+benchmark can be found in the
+[gist](https://gist.github.com/surma/0eb306fa9acc8bdf2f58150b2f1e82b4).
+
+#### C and Emscripten
+
+Emscripten is a C compiler for the WebAssembly target. Emscripten's goal is to
+function as a drop-in replacement for well-known C compilers like GCC or clang
+and is mostly flag compatible. This is a core part of the Emscripten's mission
+as it wants to make compiling existing C and C++ code to WebAssembly as easy as
+possible.
+
+Accessing raw memory is in the very nature of C and pointers exist for that very
+reason:
+
+    uint8_t* ptr = (uint8_t*)0x124;
+    ptr[0] = 0xFF;
+
+Here we are turning the number `0x124` into a pointer to unsigned, 8-bit
+integers (or bytes). This effectively turns the `ptr` variable into an array
+starting at memory address `0x124`, that we can use like any other array,
+allowing us to access individual bytes for reading and writing. In our case we
+are looking at an RGBA buffer of an image that we want to re-order to achieve
+rotation. To move a pixel we actually need to move 4 consecutive bytes at once
+(one byte for each channel: R, G, B and A). To make this easier we can create an
+array of unsigned, _32-bit_ integers. By convention, our input image will start
+at address 4 and our output image will start directly after the input image
+ends:
+
+    int bpp = 4;
+    int imageSize = inputWidth * inputHeight * bpp;
+    uint32_t* inBuffer = (uint32_t*) 4;
+    uint32_t* outBuffer = (uint32_t*) (imageSize + 4);
+
+    for (int d2 = d2Start; d2 >= 0 && d2 < d2Limit; d2 += d2Advance) {
+      for (int d1 = d1Start; d1 >= 0 && d1 < d1Limit; d1 += d1Advance) {
+        int in_idx = ((d1 * d1Multiplier) + (d2 * d2Multiplier));
+        outBuffer[i] = inBuffer[in_idx];
+        i += 1;
+      }
+    }
+
+Note: The reason we chose to start at address 4 and not 0 is because address 0
+has a special meaning in many languages: It's the dreaded null pointer. While
+technically 0 is a perfectly valid address, many languages exclude 0 as a valid
+value for pointers and either throw an exception or just tumble into undefined
+behavior.
+
+After porting the entire JavaScript function to C, we can compile [the C
+file](https://gist.github.com/surma/0eb306fa9acc8bdf2f58150b2f1e82b4#file-rotate-c)
+with `emcc`:
+
+    $ emcc -O3 -s ALLOW_MEMORY_GROWTH=1 -o c.js rotate.c
+
+As always, emscripten generates a glue code file called `c.js` and a wasm module
+called `c.wasm`. Note that the wasm module gzips to only ~260 Bytes, while the
+glue code is around 3.5KB after gzip. After some fiddling, we were able to ditch
+the glue code and instantiate the WebAssembly modules with the vanilla APIs.
+This is often possible with Emscripten as long as you are not using anything
+from the C standard library.
+
+Note: We are working with the Emscripten team to make the glue code smaller or
+even non-existent at times.
+
+#### Rust
+
+Rust is a new, modern systems programming language. Amongst other things, it
```

Comment:
you may consider trying to use some of the marketing language that the new rust website uses- particularly the bit around empowering folks who may not have originally gravitated to or felt successful in other "systems" languages

Reply:
understood.